AzureCalling/Controllers/AudioDeviceSelectionViewController.swift

```diff
@@ -0,0 +1,131 @@
+        view.isOpaque = false
+
+        // tapping anywhere on the view is the same as tapping cancel
+        let tap = UITapGestureRecognizer(target: self, action: #selector(dismissSelf))
+        tap.cancelsTouchesInView = false
+        view.addGestureRecognizer(tap)
+        view.isUserInteractionEnabled = true
+
+        createAudioDeviceOptions()
+        createDeviceTable()
+    }
+
+    @objc func dismissSelf(sender: NSObject) {
+        dismiss(animated: true, completion: nil)
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+
+        openDeviceTable()
+    }
+
+    func createDeviceTable() {
+        deviceTable = UITableView()
+        deviceTable.isHidden = true
+        deviceTable.allowsSelection = true
+        deviceTable.isScrollEnabled = false
+        deviceTable.layer.cornerRadius = 8
+        deviceTable.translatesAutoresizingMaskIntoConstraints = false
+        view.addSubview(deviceTable)
+        let audioDeviceCell = UINib(nibName: "BottomDrawerCellView",
+                                      bundle: nil)
+        deviceTable.register(audioDeviceCell, forCellReuseIdentifier: "BottomDrawerCellView")
+        deviceTable.dataSource = self
+        deviceTable.delegate = self
+        deviceTable.reloadData()
+
+        var deviceTableConstraints = [
+        deviceTable.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+        deviceTable.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+        deviceTable.heightAnchor.constraint(equalToConstant: 100)
+        ]
+
+        let hideConstraint = deviceTable.topAnchor.constraint(equalTo: view.bottomAnchor)
+        hideConstraint.priority = .defaultLow
+        deviceTableConstraints.append(hideConstraint)
+
+        NSLayoutConstraint.activate(deviceTableConstraints)
+    }
+
+    func openDeviceTable() {
+        deviceTable.isHidden = false
+
+        let centerYConstraint = NSLayoutConstraint(item: deviceTable!,
+                attribute: .bottom,
+                relatedBy: .equal,
+                toItem: self.view,
+                attribute: .bottom,
+                multiplier: 1,
+                constant: 0)
+        centerYConstraint.priority = .required
+        self.view.addConstraint(centerYConstraint)
+        UIView.animate(withDuration: 0.15, delay: 0, options: .beginFromCurrentState, animations: {
+            self.view.layoutIfNeeded()
+        }, completion: nil)
+    }
+
+    func createAudioDeviceOptions() {
+        let audioDeviceTypes = AudioSessionManager.getAllAudioDeviceTypes()
+        let currentAudioDeviceType = AudioSessionManager.getCurrentAudioDeviceType()
+
+        for audioDeviceType in audioDeviceTypes {
+            let audioDeviceOption = AudioDeviceOption(image: audioDeviceType.image, name: audioDeviceType.name, accessoryImage: audioDeviceType.accessoryImage, enabled: audioDeviceType == currentAudioDeviceType)
```

Comment:
Since there is an output model from AudioSessionManager already, would it be necessary to convert the `audioDeviceType` to `AudioDeviceOption` then to `bottonCellViewData`? Could it be `audioDeviceType` to `bottomCellViewData`? which we could delete `audioDeviceOption` data type

Reply:
Right, good catch!  I previously had BottomCellViewDataModel have different properties than AudioDeviceOption, but I made them the same at same point.  Removing